Shop

JonaLam opened the pull request:
# Description

Added a shop system
new shop scene which sells item

## Related issue(s)

#119 

## List of changes

- Changed the shop scene into something.
- Added shop item class which we place in the scene to sell the diffrent items

## Additional notes

How items are priced is not implemented in any meaningful way other than a set number on each of the shop items
the pool of items the shop can sell is just strings to their position in the files 
The inventory might make you unable to click anything, to fix you can make the inventory be a child of the inventorynode rather than the tree

Turtyo commented:
I think we should remove the test on the naming convention, it's more bothersome than anything else

Turtyo commented on Cards/CardWorld.gd:
```diff
@@ -24,6 +24,13 @@ func init_card(in_card_data: CardBase) -> void:
 
+## takes money and increases price of shop removal, should only be used when this is wanted
+func remove_this_from_player_deck_with_price() -> void:
+	ShopManager.remove_card_from_player_deck_with_price(card_data)
```
there is no check if we actually have the gold to do that right ?

Turtyo commented on Shop/shop_item_relic.gd:
```diff
@@ -0,0 +1,16 @@
+@export var clickHandler : ClickHandler
+
+func _ready() -> void:
+	var rand : RandomNumberGenerator = RandomNumberGenerator.new()
```
same

Turtyo commented on Shop/shop_item_torch.gd:
```diff
@@ -0,0 +1,33 @@
+extends ShopItem
+
+@export var clickHandler : ClickHandler
+var torchTexture : PackedScene = preload("res://Shop/tourchDisplay.tscn")
```
why is this not `@export` ? also it's `tourchDisplay` it should be renamed `torchDisplay` :laughing:

Turtyo commented on Shop/shop_item_torch.gd:
```diff
@@ -0,0 +1,33 @@
+	
+	var startPos : Vector2 = startPosTransform.position
+	
+	for i in amount:
+		var instance : TextureRect = torchTexture.instantiate()
+		heldTorches.append(instance) 
+		instance.position = startPos
+		startPos += Vector2.RIGHT * 5
+		add_child(instance)
```
a little comment to explain what you are doing here would be good

Turtyo commented on #Scenes/Events/shop/CardRemoval.gd:
```diff
@@ -0,0 +1,22 @@
+extends CardPileUISetter
```
This should be inside `SceneScripts` I think

Turtyo reviewed (changes requested):
Checking the scene inside the game, the overlay on the top right is not correctly on screen. The `proceed` button should automatically open the map, or the map movement should be set to allowed directly when entering the shop (I think the second option is better)

Opening the inventory still prevents it from interacting with anything else or being closed, as mentioned in #136 
That makes it impossible for me to add gold and thus to test that buying things in the shop actually work, how did you test that btw ?

Tysterman74 commented on Shop/ShopItem.gd:
```diff
@@ -0,0 +1,23 @@
+		InventoryManager.gold_component.lose_gold(cost)
+		_give_player_item()
+	else:
+		print("can't afford")
```
I'd say send something to the scene here to inform the user that they don't have enough money to purchase the item. Probably a debug variable in DEBUG_VAR to let us purchase w/o having to have gold, for testing purposes!

JonaLam replied:
I'll make a debug function which gives a lot of gold instead, as that is easier and should do basicly the same thing

Tysterman74 commented:
Also, if we're removing out items and consumables should we have this PR updated to remove those sections?

JonaLam commented on #Scenes/Events/shop/0.t1104.tmp:
```diff
@@ -0,0 +1,37 @@
+[gd_scene load_steps=8 format=3 uid="uid://b8hbst1klk5ng"]
+
+[ext_resource type="PackedScene" uid="uid://bam77cwf4emyr" path="res://#Scenes/TopBarOverlay.tscn" id="1_7gwvp"]
+[ext_resource type="Script" path="res://#Scenes/SceneScripts/ShopScene.gd" id="1_u1m6k"]
+[ext_resource type="Resource" uid="uid://uv2rili0xf3x" path="res://Cards/Resource/Card_Damage_EVERYTHING.tres" id="2_u60ec"]
+[ext_resource type="Texture2D" uid="uid://bo0j1ve4n2ni7" path="res://Art/Background/Shop_Background.png" id="3_6osfu"]
+[ext_resource type="Resource" uid="uid://d4lugn62mmlep" path="res://Cards/Resource/Card_DrawCards.tres" id="3_trwis"]
+[ext_resource type="PackedScene" uid="uid://cdt4k2reymt2i" path="res://Shop/shop_item.tscn" id="4_bbct0"]
+[ext_resource type="Resource" uid="uid://0x385c3nuq8f" path="res://Cards/Resource/Card_DamageAllEnemies.tres" id="4_e4vp7"]
+
+[node name="Control" type="Control"]
+layout_mode = 3
+anchors_preset = 15
+anchor_right = 1.0
+anchor_bottom = 1.0
+grow_horizontal = 2
+grow_vertical = 2
+script = ExtResource("1_u1m6k")
+cardPool = Array[Object]([ExtResource("2_u60ec"), ExtResource("3_trwis"), ExtResource("4_e4vp7")])
+
+[node name="TopBarOverlay" parent="." instance=ExtResource("1_7gwvp")]
+layout_mode = 1
+
+[node name="TextureRect" type="TextureRect" parent="."]
+layout_mode = 0
+offset_left = 20.0
+offset_top = 16.0
+offset_right = 1068.0
+offset_bottom = 1064.0
+texture = ExtResource("3_6osfu")
+expand_mode = 1
+
+[node name="ShopItem" parent="." instance=ExtResource("4_bbct0")]
+offset_left = 216.0
+offset_top = 252.0
+offset_right = 216.0
+offset_bottom = 252.0
```
Is there a good way to revert this?

Tysterman74 replied:
Here's a Stackoverflow on how to revert the file: https://stackoverflow.com/questions/18357511/git-remove-committed-file-after-push

But honestly... it's a .tmp file so I feel you could just delete this and push up that change. But I would sanity check that the shop scene is still working first after you delete and before you push up that change!

Turtyo replied:
You can just delete the file and push the change, no one else uses the tmp file (as ty said, just check that it still works correctly before pushing, but I removed tmp file in another PR and it worked fine)

JonaLam commented on Managers/ShopManager.gd:
```diff
@@ -0,0 +1,27 @@
+## or potentionaly a specific event
+func remove_card_from_player_deck_with_price(card : CardBase) -> void:
+	CardManager.current_deck.erase(card)
+	InventoryManager.gold_component.lose_gold(card_removal_price)
```
There is definitly something stoping this from being done somewhere, but I'll add checks in these areas as well just to be safe

JonaLam commented on Shop/ShopItemCard.gd:
```diff
@@ -0,0 +1,16 @@
+var card : CardBase
+
+func _ready() -> void:
+	var rand : RandomNumberGenerator = RandomNumberGenerator.new()
```
I won't implement how this wil work with seeds, as I don't really know how it works, but I'll make a placeholder function for the time being

Turtyo commented:
Yes, since we won't be using items and consumables, I would comment out / disable the items in this PR; the shop is still useful to buy torches and cards

Turtyo reviewed (changes requested):
![image](https://github.com/user-attachments/assets/a5d25da5-788b-4a56-b81c-314169e733b1)
![image](https://github.com/user-attachments/assets/fcb1ac3d-e2e3-4dfe-acff-369f84392d09)
The overlay is not positioned correctly.

The proceed button doesn't open the map.

JonaLam commented:
I removed the items and consumables from the shop, they're just not there, but I don't really see the point of deleting the files so I think it's fine to let them be there.

I also removed the proceed button and made it so you can move by opening the map.
